### PR TITLE
Generate `List-Id` header from `MessageBuilder.reference`.

### LINKED ISSUE #2461: Add List-ID header for email messages
See: https://github.com/blog/695-easily-filter-notifications-by-repository

### PR DESCRIPTION
**Pull request by tkaemming**
Adds `List-Id` header to messages built with `MessageBuilder` that
include a reference to a `Project`, `Group` or `Event` instance. This also
adds a `mail.list-namespace` option that can be used to define the RFC
2919 list namespace.

Fixes GH-2461.

### CODE REVIEW

File: src/sentry/utils/email.py [hunk 1]
```diff
@@ -112,6 +114,65 @@ def make_msgid(domain):
...
+
+default_list_type_handlers = {
+    Project: attrgetter('slug', 'organization.slug'),
+    Group: attrgetter('project.slug', 'organization.slug'),
+    Event: attrgetter('project.slug', 'organization.slug'),
+}
+
+
+make_listid_from_instance = ListResolver(
+    options.get('mail.list-namespace'),
```
Comment (tkaemming) [inline src/sentry/utils/email.py]: It'd be pretty trivial to make this not require `NOSTORE`, I just didn't.

File: src/sentry/utils/email.py [hunk 2]
```diff
@@ -10,22 +10,24 @@
 import logging
 import os
 import time
-import toronado
+from email.utils import parseaddr
+from operator import attrgetter
 from random import randrange
 
+import toronado
```
Comment (dcramer) [inline src/sentry/utils/email.py]: was this from isort? wtf isort

Comment (tkaemming) [inline src/sentry/utils/email.py]: Yeah, this is how it's supposed to work from PEP 8, this a "related third party import": https://www.python.org/dev/peps/pep-0008/#imports

File: src/sentry/options/defaults.py [hunk 3]
```diff
@@ -43,3 +43,5 @@
 # symbolizer specifics
 register('dsym.llvm-symbolizer-path', type=String)
 register('dsym.cache-path', type=String, default='/tmp/sentry-dsym-cache')
+
+register('mail.list-namespace', type=String, default='localhost', flags=FLAG_NOSTORE)
```
Comment (dcramer) [inline src/sentry/options/defaults.py]: can we:

1. add doc changes in this PR
2. expose this in the mail settings admin page

**Comment by dcramer** [comment]
in general lgtm -- I didn't read the related ticket to understand goals, but no objections

File: src/sentry/options/defaults.py (hunk 3 above)
Comment (tkaemming) [inline src/sentry/options/defaults.py]: > add doc changes in this PR

Yep.

> expose this in the mail settings admin page

I'll make it support `NOSTORE` in a subsequent change.

Comment (tkaemming) [inline src/sentry/options/defaults.py]: Er, *not* require `NOSTORE`.

Comment (dcramer) [inline src/sentry/options/defaults.py]: Oh I mostly mean just show what the value is -- we dont allow you to configure any of the mail settings in the UI yet (I dont think?)

Comment (tkaemming) [inline src/sentry/options/defaults.py]: I actually have no idea how `/manage/settings/` works now. If it's trivial to add, I'll just do it in this change.

**Comment by tkaemming** [comment]
I'm also going to add it for `Activity` types here, since that should just be another type handler. I thought it was going to be more complicated before I actually looked.

**Comment by tkaemming** [comment]
This is pending GH-2773 to avoid some annoying conflicts upon merging.

File: src/sentry/utils/email.py [hunk 4]
```diff
@@ -11,23 +11,25 @@
...
 
 from django.conf import settings
-from django.core.mail import (
-    get_connection as _get_connection,
-    send_mail as _send_mail,
-    EmailMultiAlternatives,
-)
+from django.core.mail import get_connection as _get_connection
+from django.core.mail import send_mail as _send_mail
+from django.core.mail import EmailMultiAlternatives
```
Comment (tkaemming) [inline src/sentry/utils/email.py]: isort ¯\\\_(ツ)_/¯

**Pull request merged**